src/riak_kv_pb_index.erl

```diff
@@ -56,81 +56,86 @@ init() ->
-            Bucket = bucket_type(T, B),
-            {ok, Msg, {"riak_kv.index", Bucket}}
-    end.
+    #rpbindexreq{type=T, bucket=B} = Msg = riak_pb_codec:decode(Code, Bin),
+    Bucket = bucket_type(T, B),
+    {ok, Msg, {"riak_kv.index", Bucket}}.
 
 %% @doc encode/1 callback. Encodes an outgoing response message.
 encode(Message) ->
     {ok, riak_pb_codec:encode(Message)}.
 
-validate_request(#rpbindexreq{qtype=QType, key=SKey,
-                              range_min=Min, range_max=Max,
-                              term_regex=TermRe} = Req) ->
-    {ValRe, ValErr} = case TermRe of
-        undefined ->
-            {undefined, undefined};
-        _ ->
-            re:compile(TermRe)
-    end,
 
-    if
-        QType == eq andalso not is_binary(SKey) ->
-            {error, {format, "Invalid equality query ~p", [SKey]}};
-        QType == range andalso not(is_binary(Min) andalso is_binary(Max)) ->
-            {error, {format, "Invalid range query: ~p -> ~p", [Min, Max]}};
-        ValRe =:= error ->
-            {error, {format, "Invalid term regular expression ~p : ~p",
-                     [TermRe, ValErr]}};
-        true ->
-            Query = riak_index:to_index_query(query_params(Req)),
-            case Query of
-                {ok, ?KV_INDEX_Q{start_term=Start, term_regex=Re}} when is_integer(Start)
-                       andalso Re =/= undefined ->
-                    {error, "Can not use term regular expression in integer query"};
-                _ ->
-                    Query
-            end
+validate_request(#rpbindexreq{qtype = eq, key = SKey}) when not is_binary(SKey) ->
+    {error, {format, "Invalid equality query ~p", [SKey]}};
+validate_request(#rpbindexreq{qtype = range, range_min = Min, range_max = Max})
+    when not is_binary(Min); not is_binary(Max) ->
+    {error, {format, "Invalid range query: ~p -> ~p", [Min, Max]}};
+validate_request(#rpbindexreq{term_regex = TermRe} = Req) ->
+    {ValRe, ValErr} = ensure_compiled_re(TermRe),
+    case ValRe of
+      error ->
+          {error, {format, "Invalid term regular expression ~p : ~p",
+                   [TermRe, ValErr]}};
+      _ ->
+        validate_query(Req)
     end.
 
+validate_query(Req) ->
+  Query = riak_index:to_index_query(query_params(Req)),
+  case Query of
+    {ok, ?KV_INDEX_Q{start_term = Start, term_regex = Re}} when is_integer(Start)
+                                                                andalso Re =/= undefined ->
+      {error, "Can not use term regular expression in integer query"};
+    _ ->
+      Query
+  end.
+
+ensure_compiled_re(TermRe) ->
+  case TermRe of
+      undefined ->
+        {undefined, undefined};
+      _ ->
+        re:compile(TermRe)
+  end.
+
 %% @doc process/2 callback. Handles an incoming request message.
 process(#rpbindexreq{stream = S} = Req, State) ->
     Class = case S of
         true ->
             {riak_kv, stream_secondary_index};
-        _ ->
+        false ->
```

Comment:
Missed this during my first sweep earlier. This change makes me a bit nervous. Are we positive that this value will only ever be `true` or `false`? Could it be, say, `undefined`?